terra-rust-wallet: compile errors due to lack of Sync

I believe this is related exclusively to error handling code. I'm seeing error messages such as the following:

```
error[E0277]: `*const i8` cannot be shared between threads safely
  --> terra-rust-wallet/src/lib.rs:62:72
   |
62 |         keyring.set_password(&serde_json::to_string(&wallet_internal)?)?;
   |                                                                        ^ `*const i8` cannot be shared between threads safely
   |
   = help: within `KeyringError`, the trait `Sync` is not implemented for `*const i8`
   = note: required because it appears within the type `dbus::ffi::DBusError`
   = note: required because it appears within the type `dbus::Error`
   = note: required because it appears within the type `secret_service::error::SsError`
   = note: required because it appears within the type `KeyringError`
   = note: required because of the requirements on the impl of `From<KeyringError>` for `anyhow::Error`
   = note: required by `std::convert::From::from`

error[E0277]: `*const std::ffi::c_void` cannot be shared between threads safely
  --> terra-rust-wallet/src/lib.rs:62:72
   |
62 |         keyring.set_password(&serde_json::to_string(&wallet_internal)?)?;
   |                                                                        ^ `*const std::ffi::c_void` cannot be shared between threads safely
   |
   = help: within `KeyringError`, the trait `Sync` is not implemented for `*const std::ffi::c_void`
   = note: required because it appears within the type `dbus::ffi::DBusError`
   = note: required because it appears within the type `dbus::Error`
   = note: required because it appears within the type `secret_service::error::SsError`
   = note: required because it appears within the type `KeyringError`
   = note: required because of the requirements on the impl of `From<KeyringError>` for `anyhow::Error`
   = note: required by `std::convert::From::from`
```

I'll try to investigate this a bit further myself.

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -1,0 +1,62 @@
+on:
+  push:
+    branches: [main]
+  pull_request:
+
+name: Continuous integration
+
+jobs:
+  ci:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-latest
+          - macos-latest
+          - windows-latest
+        rust:
+          - stable
+          - nightly
+          - 1.53.0
+
+    steps:
+      - name: Install deps
+        if: matrix.os == 'ubuntu-latest'
+        run: sudo apt-get install -y libdbus-1-dev
+
+      - uses: actions/checkout@v2
+
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: ${{ matrix.rust }}
+          override: true
+          components: rustfmt, clippy
+
+      - uses: Swatinem/rust-cache@v1
+        with:
+          key: ${{ runner.os }}-${{ matrix.rust }}-${{ hashFiles('Cargo.lock') }}
+
+      - uses: actions-rs/cargo@v1
+        name: Build
+        with:
+          command: build
+
+      - uses: actions-rs/cargo@v1
+        name: Test
+        with:
+          command: test
+
+      - uses: actions-rs/cargo@v1
+        name: Format
+        with:
+          command: fmt
+          args: --all -- --check
+
+      - uses: actions-rs/cargo@v1
+        name: Clippy
+        if: matrix.rust == 'stable'
+        with:
+          command: clippy
+          args: -- -D warnings

--- a/terra-rust-api/src/client.rs
+++ b/terra-rust-api/src/client.rs
@@ -286,7 +286,7 @@ impl<'a> Terra<'a> {
         }
     }
     /// helper function to generate a 'StdSignMsg' & 'Signature' blocks to be used to broadcast a transaction
-
+    #[allow(clippy::too_many_arguments)]
     fn generate_transaction_to_broadcast_fees(
         chain_id: String,
         account_number: u64,

--- a/terra-rust-wallet/src/errors.rs
+++ b/terra-rust-wallet/src/errors.rs
@@ -44,16 +44,31 @@ error_chain! {
 #[derive(Error, Debug)]
 pub enum TerraRustWalletError {
     #[error(transparent)]
-    KeyringError(#[from] ::keyring::KeyringError),
+    KeyringError(#[from] KeyringErrorAdapter),
     #[error(transparent)]
     SerdeJsonError(#[from] ::serde_json::Error),
 
     #[error("Terra Wallet `{key}` Key not found Error")]
     KeyNotFound {
         key: String,
-        source: keyring::KeyringError,
+        source: KeyringErrorAdapter,
     },
 
     #[error("unknown Terra-Rust Wallet error")]
     Unknown,
+}
+
+/// Workaround type to provide [Sync] on Linux.
+///
+/// On Linux, [keyring::KeyringError] does not implement `Sync` due to depending
+/// on an older version of the `dbus` crate. This prevents usage of `anyhow`. This
+/// wrapper is used to bypass that issue on Linux.
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub struct KeyringErrorAdapter(anyhow::Error);
+
+impl From<keyring::KeyringError> for KeyringErrorAdapter {
+    fn from(e: keyring::KeyringError) -> Self {
+        KeyringErrorAdapter(anyhow::anyhow!("Keyring error: {:?}", e))
+    }
 }

--- a/terra-rust-wallet/src/lib.rs
+++ b/terra-rust-wallet/src/lib.rs
@@ -24,6 +24,7 @@
 /// Error Messages
 pub mod errors;
 
+use crate::errors::KeyringErrorAdapter;
 //#[macro_use]
 //extern crate error_chain;
 use crate::errors::TerraRustWalletError;
@@ -59,7 +60,9 @@ impl<'a> Wallet<'a> {
         let wallet_list_name = &wallet.full_list_name();
         let keyring = keyring::Keyring::new(wallet_name, wallet_list_name);
         let wallet_internal = WalletInternal { keys: vec![] };
-        keyring.set_password(&serde_json::to_string(&wallet_internal)?)?;
+        keyring
+            .set_password(&serde_json::to_string(&wallet_internal)?)
+            .map_err(KeyringErrorAdapter::from)?;
         let string_key_name: String = String::from(wallet_name);
 
         match Wallet::get_wallets() {
@@ -98,7 +101,7 @@ impl<'a> Wallet<'a> {
     ) -> anyhow::Result<PrivateKey> {
         let full_key_name = self.full_key_name(key_name);
         let keyring = keyring::Keyring::new(&self.name, &full_key_name);
-        let phrase = &keyring.get_password()?;
+        let phrase = &keyring.get_password().map_err(KeyringErrorAdapter::from)?;
 
         match seed {
             None => Ok(PrivateKey::from_words(secp, phrase)?),
@@ -122,7 +125,9 @@ impl<'a> Wallet<'a> {
         let full_key_name = self.full_key_name(key_name);
 
         let keyring = keyring::Keyring::new(&self.name, &full_key_name);
-        keyring.set_password(pk.words().unwrap())?;
+        keyring
+            .set_password(pk.words().unwrap())
+            .map_err(KeyringErrorAdapter::from)?;
         let old_list = self.get_keys()?;
         let string_key_name: String = String::from(key_name);
         let mut new_list: Vec<String> = vec![];
@@ -142,7 +147,9 @@ impl<'a> Wallet<'a> {
     pub fn delete_key(&self, key_name: &str) -> anyhow::Result<bool> {
         let full_key_name = self.full_key_name(key_name);
         let keyring = keyring::Keyring::new(&self.name, &full_key_name);
-        keyring.delete_password()?;
+        keyring
+            .delete_password()
+            .map_err(KeyringErrorAdapter::from)?;
         let old_list = self.get_keys()?;
         let mut new_list = vec![];
         for s in old_list {
@@ -168,7 +175,9 @@ impl<'a> Wallet<'a> {
         }
         let wallet_list_name = self.full_list_name();
         let keyring = keyring::Keyring::new(&self.name, &wallet_list_name);
-        keyring.delete_password()?;
+        keyring
+            .delete_password()
+            .map_err(KeyringErrorAdapter::from)?;
         let old_list = Wallet::get_wallets()?;
         // let string_key_name: String = String::from(self.name);
         let mut new_list: Vec<String> = vec![];
@@ -202,7 +211,7 @@ impl<'a> Wallet<'a> {
             .get_password()
             .map_err(|source| TerraRustWalletError::KeyNotFound {
                 key: wallet_list_name,
-                source,
+                source: KeyringErrorAdapter::from(source),
             })?;
 
         let wallet_internal: WalletInternal = serde_json::from_str(&pass)?;
@@ -214,7 +223,8 @@ impl<'a> Wallet<'a> {
         let wallet_list_name = Wallet::wallet_list_name();
         let keyring = keyring::Keyring::new(&wallet_list_name, "wallets");
 
-        let wallet_internal: WalletListInternal = serde_json::from_str(&keyring.get_password()?)?;
+        let wallet_internal: WalletListInternal =
+            serde_json::from_str(&keyring.get_password().map_err(KeyringErrorAdapter::from)?)?;
         Ok(wallet_internal.wallets)
     }
 
@@ -223,7 +233,9 @@ impl<'a> Wallet<'a> {
         let wallet_list_name = self.full_list_name();
         let keyring = keyring::Keyring::new(&self.name, &wallet_list_name);
 
-        keyring.set_password(&serde_json::to_string(int)?)?;
+        keyring
+            .set_password(&serde_json::to_string(int)?)
+            .map_err(KeyringErrorAdapter::from)?;
         Ok(())
     }
     /// update list of wallets
@@ -231,7 +243,9 @@ impl<'a> Wallet<'a> {
         let wallet_list_name = Wallet::wallet_list_name();
         let keyring = keyring::Keyring::new(&wallet_list_name, "wallets");
 
-        keyring.set_password(&serde_json::to_string(int)?)?;
+        keyring
+            .set_password(&serde_json::to_string(int)?)
+            .map_err(KeyringErrorAdapter::from)?;
         Ok(())
     }
 }

